types/instance.d.ts

```diff
@@ -0,0 +1,109 @@
+import * as http from 'http'
+import * as http2 from 'http2'
+import * as https from 'https'
+
+import { InjectOptions, InjectPayload } from 'light-my-request'
+import { RouteOptions, RouteShorthandMethod } from './route'
+import { FastifySchema, FastifySchemaCompiler } from './schema'
+import { RawServerBase, RawRequestBase, RawReplyBase, RawServerDefault, RawRequestDefault, RawReplyDefault } from './utils'
+import { FastifyLoggerOptions } from './logger'
+import { FastifyRegister } from './register'
+import { onCloseHook, onRouteHook, onRequestHook, onSendHook, onErrorHook, preHandlerHook, preParsingHook, preSerializationHook, preValidationHook } from './hooks'
+import { FastifyRequest } from './request'
+import { FastifyReply } from './reply'
+import { FastifyError } from './error'
+import { addContentTypeParser, hasContentTypeParser } from './content-type-parser';
+
+export interface FastifyInstance<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> {
+  server: RawServer
+  prefix: string
+  log: FastifyLoggerOptions
+
+  addSchema(schema: FastifySchema): FastifyInstance<RawServer, RawRequest, RawReply>
+
+  after(err: Error): FastifyInstance<RawServer, RawRequest, RawReply>
+
+  close(closeListener?: () => void): void
+  close<T>(): Promise<T> // what is this use case? Not documented on Server#close
+
+  // should be able to define something useful with the decorator getter/setter pattern using Generics to enfore the users function returns what they expect it to
+  decorate(property: string, value: any, dependencies?: Array<string>): FastifyInstance<RawServer, RawRequest, RawReply>
```

Comment:
Hi, @Ethan-Arrowood .
I think in this case it is worth to add `symbol` type alongside with `string`.  It could be used to pass some "private" properties to request that's actually not needed for plugin users. For example @mcollina uses symbols in his graphql module https://github.com/mcollina/fastify-gql/blob/master/index.js#L93. Or in my plugin `fastify-metrics` it could be used to pass metrics property to reply for tracking request timings.